### PR TITLE
:book: describe need of RBAC permissions to allow Prometheus Operator to have access to the metrics

### LINKED ISSUE #2646: Unable to visually metrics in prometheus following steps provided
### What broke? What's expected?

After following https://book.kubebuilder.io/reference/metrics.html, the metrics from the controller don't show up in the prometheus UI.

I can see metrics in the namespace, but they come from the control plane (apiserver, kubelet, kcm, etc)

By default, the kube-prometheus installation can only monitor default, kube-system and monitoring namespaces. The solution is to add an extra Role and RoleBinding for your namespace, similar to these ones: https://github.com/prometheus-operator/kube-prometheus/blob/main/manifests/prometheus-roleSpecificNamespaces.yaml

The Role and RoleBinding could be included in the scaffolding.

### Reproducing this issue

In a brand new kubebuilder project, follow the steps in https://book.kubebuilder.io/reference/metrics.html using the latest kube-prometheus (v0.10.0)

### KubeBuilder (CLI) Version

3.3.0

### PROJECT version

3

### Plugin versions

_No response_

### Other versions

_No response_

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by machadovilaca**
Fixes #2646.

Right now, when following the [documentation](https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus) to export metrics for Prometheus, only a `ServiceMonitor` is created, but Prometheus does not have permission to access this resource. 

To fix this issue, this PR adds a `ClusterRole` and a `ClusterRoleBinding` for Prometheus `ServiceAccount` to be able to access the needed resources.

### CODE REVIEW

File: pkg/plugins/common/kustomize/v1/scaffolds/init.go
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v1/scaffolds/init.go:80]: IMO it would be more an RFE and not issue.
Thereby the title would be like:

> `:warning: Add RBAC permissions to allow Prometheus Operator have access to the metrics`

File: pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role.go
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role.go:41]: How can we check the issue + that it solves the scenario in the e2e tests: See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L255-L260
?

File: testdata/project-v3-multigroup/config/prometheus/role_binding.yaml
Comment (machadovilaca) [reply testdata/project-v3-multigroup/config/prometheus/role_binding.yaml:22]: can try to trim it down, but it is based on prometheus-operator RBAC [prometheus-operator/prometheus-operator/Documentation/rbac.md](https://github.com/prometheus-operator/prometheus-operator/blob/862598d9cdec60fd1a55efadfff0eee2cbe9f38d/Documentation/rbac.md#prometheus-rbac)

File: pkg/plugins/common/kustomize/v1/scaffolds/init.go
Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v1/scaffolds/init.go:86]: I think when a user following https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus explicitly uncomments the line `- ../prometheus` in the `config/default/kustomization.yaml`, the Role and RoleBinding should also be created alongside the ServiceMonitor

File: pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role.go
Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role.go:41]: I took a look at the tests and I think instead of just testing that we can access the service `/metrics` endpoint, https://github.com/prometheus-operator/kube-prometheus should be deployed instead of the Prometheus Operator, and then we should validate that the results of a request to `http://prometheus-k8s.monitoring.svc:9090/api/v1/targets` should contain the service monitor

File: pkg/plugins/common/kustomize/v1/scaffolds/init.go
Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v1/scaffolds/init.go:80]: changed

File: pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role_binding.go
Comment (camilamacedo86) [reply pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role_binding.go:37]: @machadovilaca, 

That is perfect. So, could we please update the metrics doc with the info? Example (something like):

<aside class="note">

Note that you need to ensure that the RBAC roles scaffold in the config/prometheus were also applied to the namespace where the manager is deployed. Otherwise, the [Prometheus operator]() will be unable to have access to the ServiceMonitor that will export the metrics. More info about the RBAC requirements when we are using the Prometheus Operator to have a Prometheus instance available and installed can be found in its [doc](link)

 </aside>

File: pkg/plugins/common/kustomize/v1/scaffolds/init.go
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v1/scaffolds/init.go:83]:
```suggestion
                // The following roles are required to allow the Promethues Operator be able to have access to
                // the ServiceMonitor. More info: https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus  
		&prometheus.Role{},
```

File: pkg/plugins/common/kustomize/v2-alpha/scaffolds/init.go
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v2-alpha/scaffolds/init.go:83]:
```suggestion
                // The following roles are required to allow the Promethues Operator be able to have access to
                // the ServiceMonitor. More info: https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus  
		&prometheus.Role{},
```

File: testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml
Comment (camilamacedo86) [inline testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml:1]: Could we also update this one by adding info on the top? something like:

# These kustomization manifest built the required manifests to allow scrapping the metrics for
# the Prometheus when/if we are using Prometheus Operator. Note that inside of this config
# we can find the ServiceMonitor which configures the metrics endpoint and the rules required 
# to allow the Prometheus Operator have access to the service monitor

File: testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml
Comment (camilamacedo86) [inline testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml:3]: could be cluster_role and cluster_role_binding instead?

**Review by camilamacedo86 — COMMENTED** [review]
Just a few nits/improvements for better clarification and maintainability.
Otherwise, it shows great 🥇 Terrific work.

**Comment by camilamacedo86** [comment]
@machadovilaca, 
Note that you need to run `make generate` to update the scaffolds before pushing the PR. 

 /ok-to-test

File: pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role_binding.go
Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/role_binding.go]: added

File: pkg/plugins/common/kustomize/v1/scaffolds/init.go
Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v1/scaffolds/init.go:85]: added

File: pkg/plugins/common/kustomize/v2-alpha/scaffolds/init.go
Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/init.go:85]: added

File: testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml
Comment (machadovilaca) [reply testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml:1]: added

**Comment by machadovilaca** [comment]
Thank you @camilamacedo86
I've updated the PR according to your comments

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:71]:
```suggestion
<aside class="note">
<h2>RBAC for Prometheus</h1>
```

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:80]:
```suggestion
</aside>

```

File: testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml
Comment (camilamacedo86) [inline testdata/project-v4-with-deploy-image/config/prometheus/kustomization.yaml:5]:
```suggestion
# This kustomization file config the required manifests to allow scrapping the metrics for
# the Prometheus using the Prometheus Operator. More info: https://github.com/prometheus-operator/prometheus-operator
# Note that inside of this config you can find the ServiceMonitor which configures the metrics endpoint 
# and the permissions (RBAC) to allow Prometheus Operator have access to the ServiceMonitor
```

File: docs/book/src/reference/metrics.md
Comment (machadovilaca) [reply docs/book/src/reference/metrics.md:71]: added

File: docs/book/src/reference/metrics.md
Comment (machadovilaca) [reply docs/book/src/reference/metrics.md:84]: added

**Review by camilamacedo86 — COMMENTED** [review]
Before merging, we need to squash the commits. Could you do so so we are able to move forward here?

File: pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/kustomization.go
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/kustomization.go:51]: Add the RBAC here is broken the kustomize build: 

> running: /home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/bin/kustomize build config/default
warning: /home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/bin/kustomize build config/default failed with error: (exit status 1) 2022/07/19 15:18:30 evalsymlink failure on '/home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/config/prometheus/cluster_role.yaml' : lstat /home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/config/prometheus/cluster_role.yaml: no such file or directory
Error: accumulating resources: recursed accumulation of path '/home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/config/prometheus': accumulating resources: accumulating resources from 'cluster_role.yaml': evalsymlink failure on '/home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/config/prometheus/cluster_role.yaml' : lstat /home/prow/go/src/sigs.k8s.io/kubebuilder/test/e2e/v3/e2e-kssw/config/prometheus/cluster_role.yaml: no such file or directory


You can test it locally to see where / how we can add these RBAC.
We might need to scaffold these ones in the config/rbac instead and then with the name prometheus_clsuter_role. 

@machadovilaca

Comment (machadovilaca) [reply pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/kustomization.go:51]: I think I fixed the issue, but `pull-kubebuilder-e2e-k8s-1-19-16 ` failed again. I think it might have been some flakiness. wdyt? Can you rerun the test?

@camilamacedo86

**Comment by joelanford** [comment]
I'm concerned with the implications that the changes from this PR will have on clusters. If I understand correctly it would mean that any operator using this new scaffolding would change the RBAC permissions of a completely unrelated operator (prometheus-operator). That's a major security problem, IMO.

A cluster admin installing the an operator from this scaffolding would be _very_ surprised if that operator delivered RBAC for a service account that was not also included with that operator.

There's also the separate problem that the scaffolding assumes a certain namespace and service account name for the prometheus operator. This is a logistical problem (how would this operator know in advance where the prometheus-operator service account is) and a security problem (how would this operator know that that service account is _actually_ the one being used by prometheus-operator; a malicious actor might be able to create that service account knowing that operators like this exist)

I think a better solution here would be to follow @camilamacedo86's advice from here: https://github.com/kubernetes-sigs/kubebuilder/issues/2646#issuecomment-1113963406

/hold

**Comment by machadovilaca** [comment]
Updated the PR to instead of applying those RBAC resources just add documentation information for users facing the issue to be aware of how it can be solved or just follow the official prometheus-operator documentation related to RBAC

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:80]: Do we need to add it here? is not that in the Prometheus Operator doc? 
Is not better only so that if they cange that we do not get this doc outdated?

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:81]: Do we need to add it here? is not that in the Prometheus Operator doc? 
Is not better only so that if they cange that we do not get this doc outdated? 

What about a note with "warning" (see that will render very nice on the docs)

```
<aside class="warning">
<h2>RBC required for Prometheus Operator have access to the metrics</h2>

If you have not Promethues installed on the cluster and you are using Promethues 
Operator then be aware that when using [prometheus-operator][rbac-promethues-operator] for
cluster monitoring, you should ensure the necessary RBAC roles and permissions. Otherwise,
Prometheus server will not be able to access the ServiceMonitor that describes
the set of targets to be monitored by Prometheus.

</aside>

Also, note that these are ClusterRoles so does not need to be applied on the ns of the operator.


```

Comment (sradco) [reply docs/book/src/reference/metrics.md:81]: Hi @camilamacedo86 @joelanford, I think that we should provide this as an optional tool and provide an api call for it.
In the documentation we will explain when it can be used and also enable changing the namespace and additional details that may change. The namespace and service account name would have a default, but we should add to the tool flags to update them.

Comment (camilamacedo86) [reply docs/book/src/reference/metrics.md:81]: @sradco is specific for Prometheus Operator and Prometheus Operator describes how to do that, see: https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/rbac.md

Therefore I do not think that Kubebuilder should be responsible for providing and maintaining helpers over "how to configure Prometheus Operator itself". The RBAC proposed to allow Prometheus Operator to have access to the ServiceMonitor used by each operator. However, users might have it set up already or prefer as @joelanford described allow this permission to specific SA instead.  

We here could/should highlight the "pre-requirement" to check the metrics with Prometheus Operator since we are suggesting its usage, see the metrics docs:

> Install Prometheus and Prometheus Operator. We recommend using [kube-prometheus](https://github.com/coreos/kube-prometheus#installing) in production **if you don’t have your own monitoring system**

Also, note that we are saying in the doc as well: 

> To achieve that, you can create a clusterRoleBinding to bind the clusterRole to the service account that your Prometheus server uses. 

Therefore, we just need a warning note pointing out the users to ensure that **if** they have Prometheus installed via the Operator then they need to ensure that Prometheus Operator has the required permissions RBAC to access the ServiceMonitor in the namespace where the Operator will be deployed and should link their documentation as well. ( We should not doc how Prometheus Operator works or how to config it, it is also part of the Prometheus Operator domain. )



 

Because

**Comment by machadovilaca** [comment]
I think we should try to meet in the middle of both opinions.
I think my current proposed changes would be an excellent middle ground between allowing users to quickly but non-manually fix the ServiceMonitor issue but still calling out the fact that we are adding permissions to an external ServiceAccount.

I suggest besides the `- ../prometheus`, we create a `- ../prometheus_rbac` section for the RBAC resources, which the user must intentionally act on to deploy.
In the docs (https://deploy-preview-2827--kubebuilder.netlify.app/reference/metrics.html#rbac-for-prometheus), we inform the user about what the problem is and warn them about the addition of new permissions to Prometheus ServiceAccount.

warn

File: docs/book/src/reference/metrics.md
Comment (sradco) [inline docs/book/src/reference/metrics.md:85]: By default, A ClusterRole will be created for the Prometheus `prometheus-k8s` ServiceAccount in the
`monitoring` namespace. Please ensure that the ServiceAccount exists and that it belongs to Prometheus. If the name or namespace ?of Prometheus? are different, you should change them in
`config/prometheus_rbac/cluster_role_binding.yaml` before applying.

"Please ensure that the ServiceAccount exists and that it belongs to Prometheus" - what does it mean? How do they make sure of that?

**Comment by camilamacedo86** [comment]
Hi @machadovilaca 

Regards : 

// I suggest besides the - ../prometheus, we create a - ../prometheus_rbac

I do not agree that we should do it at all. See that is about "How to configure your Prometheus instance installed on the cluster" therefore, it hurts concepts like single responsibility. 

Kubebuilder scaffold the project and only what is relevant to configure the project itself.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Hi @machadovilaca 

Regards: https://github.com/kubernetes-sigs/kubebuilder/pull/2827#issuecomment-1198057835
Kubebuilder scaffold the project and only what is relevant to configure the project itself. So my two cents here is that we should not add it. 

Therefore, to help out users we can address it via docs and link the Prometheus Operator documentation that explains that so users can evaluate and check how is the best setup in their own cases. 

Let's create a PR just for the docs so we can sort out the issue and then we raise an issue as RFE to study what are the use case scenarios and the best options. We might need to work on best practices, recommendations and etc. Could we move in this way

c/c @sradco

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:78]:
```suggestion
<aside class="warning">
<h2>If you are using Prometheus Operator ensure that you have the required permissions </h2>

If you are using [Prometheus Operator][rbac-promethues-operator] then be aware that by default, its RBAC rules are only enabled for the Default and kube-system namespaces. See
its documentation about how to configure the [kube-prometheus](https://github.com/prometheus-operator/kube-prometheus/blob/main/docs/monitoring-other-namespaces.md) to monitor another
namespaces with `.jsonnet`.

However, you can also enable it using RBAC. See the Prometheus Operator [Getting Started](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/getting-started.md#enable-rbac-rules-for-prometheus-pods) documentation to know how to enable the permissions on the namespace where the ServiceMonitor and manager are applied.
</aside>

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:98]:
```suggestion
```

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:109]:
```suggestion
```

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [reply docs/book/src/reference/metrics.md:78]: @machadovilaca,

(imo) we need to link here the Prometheus Operator documentation.
See that it will help the users know more about if they need it as well.

On top of that, `<aside class="warning">`  will create a note/CSS on the docs that can retain the attention of the users.

**Review by camilamacedo86 — APPROVED** [review]
@machadovilaca could you just rebase the commits so we are able to move forward?

Otherwise, IMHO it is good to fly:

/approved

/hold cancel

**Comment by camilamacedo86** [comment]
/test pull-kubebuilder-e2e-k8s-1-19-16

/hold cancel 

Now that is addressed via docs all agreed to move forward.

File: docs/book/src/reference/metrics.md
Comment (camilamacedo86) [inline docs/book/src/reference/metrics.md:88]: @alculquicondor WDYT?

File: docs/book/src/reference/metrics.md
Comment (sradco) [inline docs/book/src/reference/metrics.md:84]: @machadovilaca for some reason when I press the link it doesn't get me to the |"Enable RBAC rules for Prometheus pods" section, eventough your link it to it. Can we change the name of the link from "Getting Stared" to "Enable RBAC rules for Prometheus pods" ?

Comment (camilamacedo86) [reply docs/book/src/reference/metrics.md:84]: +1

Comment (machadovilaca) [reply docs/book/src/reference/metrics.md]: updated

File: docs/book/src/reference/metrics.md
Comment (alculquicondor) [inline docs/book/src/reference/metrics.md:83]:
```suggestion
Alternatively, you can give the Prometheus Operator permissions to monitor other namespaces using RBAC. See the Prometheus Operator
```

File: docs/book/src/reference/metrics.md
Comment (alculquicondor) [inline docs/book/src/reference/metrics.md:86]:
```suggestion
`ServiceMonitor` and manager exist.
```

**Comment by camilamacedo86** [comment]
/test pull-kubebuilder-e2e-k8s-1-19-16

/lgtm

Let's move forward with this one.

**Pull request merged**
